bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,467 @@
+
+    package: str
+    group: str
+    base_url: str
+    relative_url_path: str
+    symbol_id: str
+
+    @property
+    def url(self) -> str:
+        """Return the absolute url to the symbol."""
+        return "".join((self.base_url, self.relative_url_path))
+
+
+class QueueItem(NamedTuple):
+    """Contains a symbol and the BeautifulSoup object needed to parse it."""
+
+    symbol: DocItem
+    soup: BeautifulSoup
+
+    def __eq__(self, other: Union[QueueItem, DocItem]):
+        if isinstance(other, DocItem):
+            return self.symbol == other
+        return NamedTuple.__eq__(self, other)
+
+
+class CachedParser:
+    """
+    Get symbol markdown from pages with smarter caching.
+
+    DocItems are added through the `add_item` method which adds them to the `_page_symbols` dict.
+    `get_markdown` is used to fetch the markdown; when this is used for the first time on a page,
+    all of the symbols are queued to be parsed to avoid multiple web requests to the same page.
+    """
+
+    def __init__(self):
+        self._queue: List[QueueItem] = []
+        self._results = {}
+        self._page_symbols: Dict[str, List[DocItem]] = defaultdict(list)
+        self._item_events: Dict[DocItem, asyncio.Event] = {}
+        self._parse_task = None
+
+    async def get_markdown(self, client_session: ClientSession, doc_item: DocItem) -> str:
+        """
+        Get result markdown of `doc_item`.
+
+        If no symbols were fetched from `doc_item`s page before,
+        the HTML has to be fetched before parsing can be queued.
+        """
+        if (symbol := self._results.get(doc_item)) is not None:
+            return symbol
+
+        if (symbols_to_queue := self._page_symbols.get(doc_item.url)) is not None:
+            async with client_session.get(doc_item.url) as response:
+                soup = BeautifulSoup(await response.text(encoding="utf8"), "lxml")
+
+            self._queue.extend(QueueItem(symbol, soup) for symbol in symbols_to_queue)
+            del self._page_symbols[doc_item.url]
+            log.debug(f"Added symbols from {doc_item.url} to parse queue.")
+
+            if self._parse_task is None:
+                self._parse_task = asyncio.create_task(self._parse_queue())
+
+        self._move_to_front(doc_item)
+        self._item_events[doc_item] = item_event = asyncio.Event()
+        await item_event.wait()
+        return self._results[doc_item]
+
+    async def _parse_queue(self) -> None:
+        """
+        Parse all item from the queue, setting associated events for symbols if present.
+
+        The coroutine will run as long as the queue is not empty, resetting `self._parse_task` to None when finished.
+        """
+        log.trace("Starting queue parsing.")
+        while self._queue:
+            item, soup = self._queue.pop()
+            try:
+                markdown = get_symbol_markdown(soup, item)
+                await doc_cache.set_if_exists(item, markdown)
```

Comment:
Only user fetched symbols go to redis as pages may contain a lot of symbols (monolithic doc pages like discordpy's api reference may contain thousands behind a single url) that simply won't ever get fetched by user or very rarely, storing them feels like a waste when there's only a small cost to the fetching and in case it's a more used url it'll already be cached in CachedParser's internal dict as mentioned in https://github.com/python-discord/bot/pull/1014#discussion_r532129908

Reply:
Is it really a small cost? You'd end up fetching the page and parsing all its symbols, which is especially bad in monolithic pages as you mentioned. Every time the bot restarts, the dict is cleared. It's conceivable that for every restart, a different symbol from the same page would be requested. Though individual symbols may be rare, combined together, the chance of _any_ "rare"/new symbol being request seems relatively high.